### PR TITLE
New Provision: Part 1

### PR DESCRIPTION
**Pull request by chrisnojima**
This is the small changes that are required for the new Provisioning flow

I pull out the native nav fixes and engine/saga fixes into sep. prs against master so those can go in early

This won't pass CI or work without part 2

### CODE REVIEW

File: shared/__mocks__/react-redux.js [hunk 1]
```diff
@@ -17,12 +17,23 @@ const selectorDelegatorFactory = (dispatch, options) => {
   return (state, ownProps) => {
     try {
       // call the factory under the display name to get the view props
-      const viewProps = state[name](ownProps)
-      // replaces what is usually the output of mergeProps
-      return viewProps
+      const mapper = state[name]
```
Comment (chrisnojima) [inline shared/__mocks__/react-redux.js]: more detailed errors.

File: shared/actions/config/index.js [hunk 2]
```diff
@@ -143,6 +142,7 @@ const bootstrap = (opts: $PropertyType<ConfigGen.BootstrapPayload, 'payload'>):
         if (!didInitialNav) {
           dispatch(async () => {
             if (getState().config.loggedIn) {
+              dispatch(LoginGen.createNavBasedOnLoginAndInitialState())
```
Comment (chrisnojima) [inline shared/actions/config/index.js]: i do want to clean up how this works but short term we need to call this more often sometimes

File: shared/actions/json/config.json [hunk 3]
```diff
@@ -7,6 +7,9 @@
     "dumpLogs": {
       "reason": ["'quitting through menu'"]
     },
+    "openAppSettings": {
```
Comment (chrisnojima) [inline shared/actions/json/config.json]: now an action and not a function we call. helps with storybooks also

File: shared/actions/pinentry.js [hunk 4]
```diff
@@ -54,7 +54,7 @@ function _onNewPinentry(action: PinentryGen.NewPinentryPayload) {
 
 function _onSubmit(action: PinentryGen.OnSubmitPayload) {
   const {sessionID, passphrase} = action.payload
-  _respond(sessionID, {passphrase, storeSecret: false})
+  _respond(sessionID, {passphrase})
```
Comment (chrisnojima) [inline shared/actions/pinentry.js]: we assume its false always

File: shared/actions/platform-specific.native.js [hunk 5]
```diff
@@ -259,8 +259,19 @@ function configurePush() {
   })
 }
 
-function openAppSettings() {
-  Linking.openURL('app-settings:')
+const openAppSettings = () => {
+  if (isAndroid) {
```
Comment (chrisnojima) [inline shared/actions/platform-specific.native.js]: supports android now, moved out from invite stuff

File: shared/actions/push.js [hunk 6]
```diff
@@ -323,8 +318,7 @@ function* pushSaga(): Saga.SagaGenerator<any, any> {
   yield Saga.safeTakeEvery(PushGen.notification, pushNotificationSaga)
   yield Saga.safeTakeEveryPure(ConfigGen.mobileAppState, resetHandledPush)
   yield Saga.safeTakeEvery(ConfigGen.mobileAppState, mobileAppStateSaga)
+  yield Saga.safeTakeEveryPureSimple(LoginGen.logout, deletePushToken)
```
Comment (chrisnojima) [inline shared/actions/push.js]: cleaning this up

File: shared/constants/engine.js [hunk 7]
```diff
@@ -32,13 +34,11 @@ function _sagaWaitingDecorator(rpcNameKey, saga, waitingAction) {
     if (waitingAction) {
       yield Saga.put(waitingAction(false))
     }
-    yield Saga.put(EngineGen.createWaitingForRpc({name: rpcNameKey, waiting: false}))
```
Comment (chrisnojima) [inline shared/constants/engine.js]: this was just used by login and its not anymore so we can entirely remove the engine store

File: shared/engine/saga.js [hunk 8]
```diff
@@ -24,16 +24,48 @@ type EmittedFinished = {
 type CallbackWithResponse = (any, CommonResponseHandler, TypedState) => ?RS.Effect | ?Generator<any, any, any>
 type CallbackNoResponse = (any, TypedState) => ?RS.Effect | ?Generator<any, any, any>
 
+// Wraps a response to update the waiting state
+const makeWaitingResponse = (r, waitingKey) => {
```
Comment (chrisnojima) [inline shared/engine/saga.js]: i have a standlone pr for this also

File: shared/package.json [hunk 9]
```diff
@@ -138,15 +138,15 @@
     "prettier": "1.12.1",
     "prop-types": "15.6.1",
     "purepack": "1.0.4",
-    "qrcode-generator": "1.3.1",
+    "qrcode-generator": "git://github.com/keybase/qrcode-generator#keybase-changes-off-140",
```
Comment (chrisnojima) [inline shared/package.json]: we have our own fork so we can color the black to blue 👍

File: shared/package.json [hunk 10]
```diff
@@ -138,15 +138,15 @@
...
     "re-reselect": "2.1.0",
     "react": "16.4.0",
     "react-dom": "16.4.0",
     "react-gateway": "3.0.0",
     "react-list": "0.8.9",
     "react-measure": "^2.0.0",
     "react-native": "git://github.com/keybase/react-native#keybase-fixes-off-543",
-    "react-native-camera": "0.10.0",
+    "react-native-camera": "1.1.4",
```
Comment (chrisnojima) [inline shared/package.json]: this was crazy old but was afraid to touch it until now. required updates to gradle (needed for  RN 0.56 anyways)

File: shared/stories/setup-app.native.js [hunk 11]
```diff
@@ -9,7 +8,8 @@ const load = () => {
 
   const StorybookUI = getStorybookUI({
     host: 'localhost',
-    onDeviceUI: true,
+    // set this to true to show the in-app UI or just use the web ui
+    onDeviceUI: false,
```
Comment (chrisnojima) [inline shared/stories/setup-app.native.js]: you can contorl rn-storybook with chrome, just hit localhost: 7007

**Comment by chrisnojima** [comment]
@keybase/react-hackers this is ready

**Comment by chrisnojima** [comment]
- [ ] Merge after release

File: shared/actions/provision.js [hunk 12]
```diff
@@ -0,0 +1,438 @@
...
+import {isMobile} from '../constants/platform'
+import HiddenString from '../util/hidden-string'
+import {type TypedState} from '../constants/reducer'
+import {niceError} from '../util/errors'
+import {devicesTab as settingsDevicesTab} from '../constants/settings'
+import type {CommonResponseHandler} from '../engine/types'
+
+const devicesRoot = isMobile ? [Tabs.settingsTab, settingsDevicesTab] : [Tabs.devicesTab]
+
+type ValidCallbacks =
```
Comment (buoyad) [inline shared/actions/provision.js]: :sheep: `ValidCallback`

File: shared/actions/provision.js [hunk 13]
```diff
@@ -0,0 +1,438 @@
...
+  | 'keybase.1.provisionUi.chooseGPGMethod'
+  | 'keybase.1.secretUi.getPassphrase'
+
+const cancelOnCallback = (_: any, response: CommonResponseHandler, __: any) => {
+  response.error({code: RPCTypes.constantsStatusCode.scgeneric, desc: Constants.cancelDesc})
+}
+const ignoreCallback = (_: any, __: any) => {}
+
+// The provisioning flow is very stateful so we use a class to handle bookkeeping
+// We only allow one manager to be alive at a time
```
Comment (buoyad) [inline shared/actions/provision.js]: Is anything guaranteeing this? You could make this an actual singleton and have a `reInit` method on it.

File: shared/actions/login.js [hunk 14]
```diff
@@ -104,615 +64,191 @@ function* navBasedOnLoginAndInitialState(): Saga.SagaGenerator<any, any> {
...
       // If the initial state is not set yet, navigate to the people
       // tab without setting state.routeTree.loggedInUserNavigated to true.
-      yield Saga.put(navigateTo([peopleTab], null, 'initial-default'))
+      yield Saga.put(RouteTree.navigateTo([Tabs.peopleTab], null, 'initial-default'))
     }
   } else if (registered) {
     // relogging in
-    yield Saga.put(switchRouteDef(loginRouteTree))
+    yield Saga.put(RouteTree.switchRouteDef(loginRouteTree))
     yield Saga.put.resolve(getExtendedStatus())
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/actions/login.js]: which call black bars? switcRouteDef?

File: shared/actions/provision.js (hunk 13 above)
Comment (chrisnojima) [reply shared/actions/provision.js]: i spent some time doing this but it makes the flow typing really crappy and complicated. This class isn't exported at all so its just used internally. I'm not very worried about someone making more of these things

Comment (chrisnojima) [reply shared/actions/provision.js]: i'm making it a little harder to make an instance of this but not making it impossible

File: shared/actions/login.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (buoyad) [reply shared/actions/login.js]: @chrisnojima `getExtendedStatus`

**Comment by chrisnojima** [comment]
@buoyad added some fixes
- [x] Added some debugging for waiting states. Opt in by setting a flag to true
- [x] Don't reset waiting state on store reset. When you log out we entirely reset the store but we don't clean up these waiting actions per se. Instead just leave the counts so we can see whats going on and not lose our state
- [x] Use WaitingConstants.anyWaiting() helper to figure out if you're waiting. Some buttons / things would be true if waiting was negative
- [x] If you're installing electron sometimes the postinstall step could mess you up. defer using electron until your actually doing start-hot

File: shared/constants/provision.js [hunk 15]
```diff
@@ -0,0 +1,56 @@
+// @flow
+import * as I from 'immutable'
+import * as DeviceTypes from './types/devices'
+import * as Types from './types/provision'
+import * as RPCTypes from './types/rpc-gen'
+import HiddenString from '../util/hidden-string'
+
+export const waitingKey = 'provision:waiting'
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/constants/provision.js]: should be fixed, lemme know if you can repro

File: shared/provision/code-page/index.js [hunk 16]
```diff
@@ -0,0 +1,480 @@
...
+          }
+        >
+          <RequireImage
+            src={this.state.tab === 'QR' ? blueBackground : greenBackground}
+            style={
+              this.props.currentDeviceAlreadyProvisioned ? styles.backgroundOnLeft : styles.backgroundOnRight
+            }
+          />
+        </Box2>
+        <BackButton
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/provision/code-page/index.js]: fixed

**Comment by chrisnojima** [comment]
@buoyad all should be fixed

**Comment by chrisnojima** [comment]
- [ ] iphone x

**Comment by chrisnojima** [comment]
@buoyad also fixed the iphone x thing which we talked about irl

**Comment by chrisnojima** [comment]
old vs new

computer to computer
- [ ] oc1nc2
- [ ] nc1oc2

phone to phone
- [ ] op1np2
- [ ] np1op2

computer to phone
- [ ] oc1np2
- [ ] nc1op2

phone to computer
- [ ] op1nc2
- [ ] np1oc2

**Pull request merged**